### PR TITLE
Add HOGP composite BO tutorial notebook

### PR DESCRIPTION
**Pull request by wjmaddox**
Summary: Adds a tutorial describing how to perform composite BO with the HOGP to botorch.

Differential Revision: D29619715

### CODE REVIEW

File: tutorials/composite_bo_with_hogp.ipynb [hunk 1]
```diff
@@ -0,0 +1,646 @@
...
+   "metadata": {},
+   "source": [
+    "## Composite Bayesian Optimization with the Higher Order Gaussian Process"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this tutorial, we're going to explore composite Bayesian optimization [Astudillo & Frazier, ICML, '19](https://proceedings.mlr.press/v97/astudillo19a.html) with the Higher Order Gaussian Process (HOGP) model of [Zhe et al, AISTATS, '19](http://proceedings.mlr.press/v89/zhe19a.html). The setup for composite Bayesian optimization is that we have a function mapping input parameters to several outputs, and a second cheaper (and differentiable) function describing the quality of the functional output. We wish to find input parameters that maximize the output metric function in a black-box manner. \n",
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:14]: I guess you don't necessarily need the composite function to be differentiable, only if you want to do gradient-based optimization. 

```suggestion
    "In this tutorial, we're going to explore composite Bayesian optimization [Astudillo & Frazier, ICML, '19](https://proceedings.mlr.press/v97/astudillo19a.html) with the Higher Order Gaussian Process (HOGP) model of [Zhe et al, AISTATS, '19](http://proceedings.mlr.press/v89/zhe19a.html). The setup for composite Bayesian optimization is that we have an unknown (black box) function mapping input parameters to several outputs, and a second, known function describing the quality of the functional output. We wish to find input parameters that maximize the output metric function. \n",
```

File: tutorials/composite_bo_with_hogp.ipynb [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+    "\n",
+    "Specifically, this can be described as $\\max_{x \\in \\mathcal{X}} g(f(x)),$ where $f: \\mathbb{R}^d \\rightarrow \\mathbb{R}^o$ and $g: \\mathbb{R}^o \\rightarrow \\mathbb{R}.$ As in traditional Bayesian optimization, we are going to construct a Gaussian process surrogate model over the expensive to evaluate function $f(.),$ and will use a HOGP to model this function. "
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:16]:
```suggestion
    "Specifically, this can be described as $\\max_{x \\in \\mathcal{X}} g(f(x)),$ where $f: \\mathbb{R}^d \\rightarrow \\mathbb{R}^o$ and $g: \\mathbb{R}^o \\rightarrow \\mathbb{R}$, where $f$ is unknown and $g$ is known. As in traditional Bayesian optimization, we are going to construct a Gaussian process surrogate model over the expensive to evaluate function $f(.),$ and will use a HOGP to model this function. "
```

File: tutorials/composite_bo_with_hogp.ipynb [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### HOGP model description\n",
+    "\n",
+    "The [Higher Order Gaussian Process (HOGP) model](https://proceedings.mlr.press/v89/zhe19a.html) is a Gaussian process model designed specifically to operate over tensors or multi-dimensional arrays and exploits significant strucutre in the tensor to be able to operate efficiently. Specifically, the HOGP takes as inputs $y \\in \\mathbb{R}^{N \\times d_2 \\times \\cdots \\times d_M}$ and assumes that $\\text{vec}(y) \\sim \\mathcal{N}(0, \\otimes_{i=1}^M K_i + \\sigma^2 I),$ where $K_1 = K_{XX}.$ Each dimension of the tensor has its own kernel function, $K_i,$ as well as a set of $d_i$ latent parameters that can be optimized over.\n",
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:25]: They call it High-order in the paper
```suggestion
    "The [High Order Gaussian Process (HOGP) model](https://proceedings.mlr.press/v89/zhe19a.html) is a Gaussian process model designed specifically to operate over tensors or multi-dimensional arrays and exploits structure in the tensor to be able to operate efficiently. Specifically, the HOGP takes as inputs $y \\in \\mathbb{R}^{N \\times d_2 \\times \\cdots \\times d_M}$ and assumes that $\\text{vec}(y) \\sim \\mathcal{N}(0, \\otimes_{i=1}^M K_i + \\sigma^2 I),$ where $K_1 = K_{XX}.$ Each dimension of the tensor has its own kernel function, $K_i,$ as well as a set of $d_i$ latent parameters that can be optimized over.\n",
```

File: tutorials/composite_bo_with_hogp.ipynb [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+    "\n",
+    "Recently, [Maddox et al, '21](https://arxiv.org/abs/2106.12997) proposed a method for computing posterior samples from the HOGP by exploiting structure in the posterior distribution, thereby enabling its usage in BO settings."
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:27]:
```suggestion
    "Recently, [Maddox et al, '21](https://arxiv.org/abs/2106.12997) proposed a method for computing posterior samples from the HOGP by exploiting structure in the posterior distribution, thereby enabling its usage in BO settings. While they show that this approach allows to use composite BO on problems with tens or thousands of outputs, for scalability we consider a much smaller example here (that does not require GPU acceleration)."
```

File: tutorials/composite_bo_with_hogp.ipynb [hunk 5]
```diff
@@ -0,0 +1,646 @@
...
+   "source": [
+    "### Problem Description"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We use a simple test problem describing the concentration of pollutants after a chemical spill from [Astudillo & Frazier, ICML, '19](https://proceedings.mlr.press/v97/astudillo19a.html) defined over a $3 \\times 4$ grid of values $s,t$ and we wish to optimize the parameters w.r.t. their true values, to estimate the true value of parameters, $x = [M, D, L, \\tau].$ The function is given by \n",
+    "$$ f(s,t | M, D, L, \\tau) := \\frac{M}{\\sqrt{4 \\pi D t}}  \\exp\\{-\\frac{s^2}{4Dt}\\} + \\frac{1_{t > \\tau} M}{\\sqrt{4 \\pi D(t - \\tau)}} \\exp\\{- \\frac{(s - L)^2}{4 D (t - \\tau)}\\}, $$\n",
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:118]: should there be some \left / \right here to make things more readable?

File: tutorials/composite_bo_with_hogp.ipynb [hunk 6]
```diff
@@ -0,0 +1,646 @@
...
+   "metadata": {},
+   "source": [
+    "### Problem Description"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We use a simple test problem describing the concentration of pollutants after a chemical spill from [Astudillo & Frazier, ICML, '19](https://proceedings.mlr.press/v97/astudillo19a.html) defined over a $3 \\times 4$ grid of values $s,t$ and we wish to optimize the parameters w.r.t. their true values, to estimate the true value of parameters, $x = [M, D, L, \\tau].$ The function is given by \n",
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:117]: maybe note here that it's important for our setup that the function is diff'able and implemented in torch so we can backprop through it.

File: tutorials/composite_bo_with_hogp.ipynb [hunk 7]
```diff
@@ -0,0 +1,646 @@
...
+    "    return cands_nlzd"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# ZEROISH = 1e-2"
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:178]: do you still need this? (usage commented out below)

File: tutorials/composite_bo_with_hogp.ipynb [hunk 8]
```diff
@@ -0,0 +1,646 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Below is a wrapped function to help us define bounds on the parameter space, we can also vary the size of the grid if we'd like to."
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:185]: should talk about how this constructs a `GenericMCObjective` and how it will be used

File: tutorials/composite_bo_with_hogp.ipynb [hunk 9]
```diff
@@ -0,0 +1,646 @@
...
+   "outputs": [],
+   "source": [
+    "from botorch.optim.fit import fit_gpytorch_torch"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As a word of caution, we've found that when fitting the HOGP model, using first-order optimizers (e.g. Adam) as is used in `fit_gpytorch_torch` tends to outperform second-order optimizers such as L-BFGS-B due to the large number of free parameters in the HOGP."
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:281]: Is the reason that L-BFGS-B fails or that it is "too good" and tends to overfit a model that might not be sufficiently regularized?

File: tutorials/composite_bo_with_hogp.ipynb [hunk 10]
```diff
@@ -0,0 +1,646 @@
...
+    "            }\n",
+    "            sampler = IIDNormalSampler(num_samples=128)\n",
+    "\n",
+    "            train_Y_ei = objective(train_Y[\"ei\"]).unsqueeze(-1)\n",
+    "            model_ei = SingleTaskGP(\n",
+    "                train_X[\"ei\"],\n",
+    "                train_Y_ei,\n",
+    "                input_transform=Normalize(train_X[\"ei\"].shape[-1]),\n",
+    "                outcome_transform=Standardize(train_Y_ei.shape[-1]),\n",
+    "            )\n",
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:485]: Should explain before this cell that you'll do the comparison against BO on the aggregate metrics and random.

File: tutorials/composite_bo_with_hogp.ipynb [hunk 11]
```diff
@@ -0,0 +1,646 @@
...
+    "            # generate qEI candidate (multi-output modeling)\n",
+    "            qEI_hogp_cf = qExpectedImprovement(\n",
+    "                model_ei_hogp_cf,\n",
+    "                best_f=best_f[\"ei_hogp_cf\"],\n",
+    "                sampler=sampler,\n",
+    "                objective=objective,\n",
+    "            )\n",
+    "            cands[\"ei_hogp_cf\"] = optimize_ei(qEI_hogp_cf, bounds, **optimize_acqf_kwargs)\n",
+    "            del model_ei_hogp_cf\n",
+    "            torch.cuda.empty_cache()\n",
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:517]: maybe leave a comment why you're doing this.

File: tutorials/composite_bo_with_hogp.ipynb [hunk 12]
```diff
@@ -0,0 +1,646 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "n_init = 30\n",
+    "n_batches = 15\n",
+    "batch_size = 4\n",
+    "n_trials = 3"
```
Comment (Balandat) [inline tutorials/composite_bo_with_hogp.ipynb:265]: Can you gate this via `os.environ.get("SMOKE_TEST")` to tiny values as done in the other tutorials? That way things can run quickly in the integration tests.

**Review by Balandat — COMMENTED** [review]
Looks like there is an issue when trying to run the tutorial on a CPU only machine? https://github.com/pytorch/botorch/pull/864/checks?check_run_id=3048279670

The results don't look terribly convincing, probably need some more replicates?

Otherwise left a number of inline comments.

File: tutorials/composite_bo_with_hogp.ipynb (hunk 3 above)
Comment (wjmaddox) [reply tutorials/composite_bo_with_hogp.ipynb:25]: We should probably change over internally as well.

File: tutorials/composite_bo_with_hogp.ipynb (hunk 9 above)
Comment (wjmaddox) [reply tutorials/composite_bo_with_hogp.ipynb:281]: That's what it was looking like.

File: tutorials/composite_bo_with_hogp.ipynb (hunk 5 above)
Comment (wjmaddox) [reply tutorials/composite_bo_with_hogp.ipynb:118]: Not sure how to add that into latex as its not showing up properly in the markdown.

File: tutorials/composite_bo_with_hogp.ipynb (hunk 9 above)
Comment (Balandat) [reply tutorials/composite_bo_with_hogp.ipynb]: I see. Maybe worth mentioning sth along those lines so his doesn't seem like magic. Adding proper regularizations sounds like a "future work" item...

**Review by Balandat — APPROVED** [review]
thanks!
